Atolcd bug fixes : Offline mode

ftavin opened the pull request:
This commits handle offline mode for contacts and LayoutGroupIteration.
The first commit also adds tomcat in docker-compose file to help dealing with offline mode dev using docker (instructions in commit message)

aponcon-atolcd committed:
Add tomcat in docker-compose file

This tomcat can automatically deploy the sigmah application when packaged
at startup.
It can be started individually by running the following command :
docker-compose up -d tomcat

Just be careful, the sigmah application will be running inside the docker container.
It means the only postgresql database available will be the one from the docker-compose file.
You can retrieve the IP address of the postgresql by running the following command :
docker-compose run postgresql env

This IP address should next be used for configuring the following file :
{project directory}/target/sigmah-2.2-SNAPSHOT/WEB-INF/classes/META-INF/persistence.xml
(file generated after generating the WAR package with Maven)

Change-Id: If9ae1e88a5811ddccb30b9c1f1fd4ecf89c0b791

ftavin committed:
Offline mode for contact

(refs #634)

Change-Id: I86dd634e4f0337aa151e14cc6c5d9b6cfa5f22eb

ftavin committed:
Offline mode for LayoutGroupIterations

(refs #501)

Change-Id: I3728fe3e0daa38c0e140f0ef970eaf08d976d275

vdelesse committed:
Bug fixes for offline mode

(refs #634 #501)

Change-Id: Ieabe0f56c7a757c7eb411de678b3acfbda25b557

osarrat commented:
Wow !
That's a great work !

While I will assess that it does functionally the job, could you improve a bit code quality as specified by this Codacy report : https://www.codacy.com/app/osarrat/sigmah/pullRequest?prid=733364 ? (mostly unused imports to remove)

numero-six commented on docker-compose.yml:
```diff
@@ -13,3 +13,11 @@ mailhog:
+  links:
+    - postgresql
+  volumes:
+    - ./target/sigmah-2.2-SNAPSHOT/:/usr/local/tomcat/webapps/sigmah/
```
Can you find a way without having the version `2.2-SNAPSHOT`?

numero-six replied:
The documentation you put in the commit message is interesting, could you put it in a comment in this file?

numero-six commented on src/main/java/org/sigmah/shared/dto/element/FlexibleElementDTO.java:
```diff
@@ -297,6 +299,11 @@ public void componentSelected(ButtonEvent ce) {
 					}
 				});
+				
+				if (ElementTypeEnum.CONTACT_LIST == this.getElementType() && Profiler.INSTANCE.getApplicationStateManager().getLastState() == ApplicationState.OFFLINE) {
```
This piece of code

```java
Profiler.INSTANCE.getApplicationStateManager().getLastState() == ApplicationState.OFFLINE
```

is copy-pasted everywhere (sometimes in a method, sometimes not). Can you factorize in a single method somewhere in `Profiler.INSTANCE` or in `ApplicationStateManager`? I guess what you are actually looking for is the flag `ApplicationStateManager#online`

numero-six commented on src/main/java/org/sigmah/client/ui/presenter/contact/ContactDetailsPresenter.java:
```diff
@@ -578,78 +579,147 @@ public void onRequiredValueChange(RequiredValueEvent event) {
-                updateCurrentContact(contactDTO, (DefaultContactFlexibleElementDTO) event.getSource(), event.getSingleValue());
-              }
-            }
+		if (!iterationChanges.isEmpty()) {
```
Too long, please break in smaller methods (or maybe extract anonymous class to inner?)

ftavin committed:
Fix details to please Codacy

Change-Id: I82d1df60a609dd8f2f1d8629a55629ab3ad77b8f

osarrat commented:
Thanks @ftavin for those last modifications. Would it be possible for you to also take into account @numero-six requested changes?

tbroyer committed:
Use docker-compose volume instead of .docker/

To remove the volume, use 'docker-compose down -v'
(instead of something like 'sudo rm -rf .docker/')

Change-Id: I3e70577be48b83083fa035a461189ee9a33764f2

tbroyer committed:
Add jetty-maven-plugin to allow developers to easily run Sigmah

To run Sigmah in an environment similar to the one
from 'mvn gwt:run', use

    mvn jetty:run-exploded -DskipTests

This is not equivalent to a production environment, but it
allows testing the offline mode while in GWT dev mode:

 * run 'mvn gwt:run -Dgwt.noserver' to start GWT dev mode
 * run 'mvn jetty:run-exploded -DskipTests -Dgwt.compiler.skip'
 * load the application and let it synchronize for offline use
 * stop Jetty, and reload: the app should now run offline and
   in GWT dev mode

Change-Id: I1c73637bb03aaa1a49d6b413e6f7d07cdb061833

tbroyer committed:
Revert "Add tomcat in docker-compose file"

An equivalent tomcat container can be run with

    docker run -it --rm \
        --link $(docker-compose ps -q postgresql):postgresql \
        -v $PWD/target/sigmah-2.2-SNAPSHOT/:/usr/local/tomcat/webapps/sigmah/:ro \
        -p 8080:8080 \
        tomcat:8.0

The host to connect to (and put in persistence.xml) will then
be 'postgresql', as given as the alias in the --link parameter.

This reverts commit 7c3e7f5f1c7e60eae7a12d5b02b4cc855550284c.

Change-Id: Ib60a265b32f340ea13a8aa67dd3c096a81d91bc4

tbroyer committed:
Cleanup: remove H2 dependency

H2 was only actually needed for a simple StringUtils.equals()
method which is equivalent to Java 7's Objects.equals()

Change-Id: Ic1c53fda65017b9ac18b12f65d89a365bacb8a28

tbroyer committed:
Cleanup: properly wrap all JS callbacks in $entry()

Change-Id: I411b922f3c05348fff2b1b9a1a9e9cd728d78cf4

tbroyer committed:
Revert use of -1 as "absent" iterationId

The change wasn't fully propagated, and actually not needed:
the root cause was (probably) that one JSNI method wasn't
prepared to handle a 'null' value.
The change in ProjectDetailsPresenter totally broke project
details when online (at least for contact list fields)

Change-Id: I5de90030a94a599a6ffd1c406eb3f523defbd27f

tbroyer committed:
Add docker-tomcat.sh to run a tomcat docker container

Change-Id: I8152cff6631486eec0bfd82d34c261354099cbdf

numero-six commented on docker-tomcat.sh:
```diff
@@ -0,0 +1,7 @@
+
+exec docker run -it --rm \
+            --link $(docker-compose ps -q postgresql):postgresql \
+            -v $PWD/target/sigmah-2.2-SNAPSHOT/:/usr/local/tomcat/webapps/sigmah/:ro \
```
Please, remove hard-coded `2.2-SNAPSHOT`.

You can start your script with something like

```bash
# gives SIGMAH_VERSION="2.2-SNAPSHOT"
SIGMAH_VERSION=`mvn --quiet -Dexec.executable='echo' -Dexec.args='${project.version}' --non-recursive exec:exec`
```

See also https://stackoverflow.com/questions/3545292/how-to-get-maven-project-version-to-the-bash-command-line

ftavin committed:
Refactorize isOfflineMode methods as one in Profiler

Change-Id: I20b4faa59f0ab68099a35479272f14a9b2260621

ftavin committed:
Split UpdateContact in ContactDetailsPresenter in smaller methods

Change-Id: If9d901eb41bd61f7a336c116446bbc557f860812

ftavin committed:
Retrieve Sigmah version using maven for docker-tomcat.sh

Change-Id: Ia5ad5ac0140f27dacc3c4085a2fa618111435c11

osarrat commented:
Hi @numero-six 
Is this PR mergeable according to you ?
Best

ghost commented:
Codacy has good points but nothing like an error.

Does it ends the 2.2 developments? If yes I would be happy to tag the final `2.2`.

I'm asking because 2.3 development has started but the roadmap show that there are remaining issues.

osarrat commented:
Thanks @numero-six ! I merge it now.
Regarding making it the final v2.2, I'll come back on this issue next week.

osarrat commented:
Hi @numero-six and @ftavin 

I wish to discuss with @ftavin on whether this v2.2-rc6 should be the final v2.2 or not.
So the point is not concluded today, and since we need to publish this v2.2-rc6 for a usage by CARE today, we can consider that a v2.2-rc6 will be released.
We will see later on if we releease this v2.2-rc6 also as the final v2.2.

Best !

ghost commented:
Hi! `2.2-rc6` tag published

osarrat commented:
Great, thank you !
